docs/cheat-sheets/java-xxe.mdx

```diff
@@ -0,0 +1,1839 @@
+---
+slug: java-xxe
+toc_max_heading_level: 4
+---
+
+# XML External Entity prevention for Java
+
+This is an XML External Entity (XXE) prevention cheat sheet by r2c. It contains code patterns of potential XXE in an application. The recommendations in this cheat sheet pave a safe road for developers that mitigate the possibility of XXE in their code. By following these recommendations, you can be reasonably sure that their code is free of XXE.
+
+## Mitigation Summary
+
+It is not easy to summarize Java XML security. The XML standard allows multiple ways to include external content, including but not limited to external Document Type Definitions (DTD), external Entity References to external data, general entity references, and external parameter references. There are also several ways to include external content in XML Schemas and stylesheets.
+
+The Java API for XML Processing (JAXP) contains multiple interfaces to process XML content:
+
+- The Document Object Model (DOM) interface
+- The Simple API for XML (SAX) interface
+- The Streaming API for XML (StAX) interface
+
+Each of these interfaces has its mitigation measures to disable the use of different sources of external content. The&nbsp;inconsistency in availability and even the effect of these measures across the different APIs, combined with the variety of ways to include external content makes it difficult to provide general mitigation recommendations. Many configurations that are secure for one parser are not secure for another.
+However, the following settings have a consistent effect on a subset of parsers
+
+- For `SAXBuilder`, `SAXParser`, and `XMLReader` use:
+    ```java
+    setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+    ```
+- For `DocumentBuilderFactory`, `TransformerFactory`, `SAXTransformer`, `SchemaFactory`, and `Validator` use:
+    ```java
+    setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    ```
+- For `SAXReader` the only secure configuration is:
+    ```java
+    setFeature("http://apache.org/xml/features/<b>disallow-doctype-decl", true);
+
+    ```
+
+The complete [Semgrep ruleset](https://semgrep.dev/r?q=java.lang.security.xxe&lang=Java) that covers vulnerabilities mentioned in this cheat sheet is available as part of the rules available in the Team tier. This ruleset includes rules to verify these settings as well as several other combinations that are secure.
```

Comment:
The list above this sentence is ways that a *single* security feature secures a parser. However, it is possible to combine several security features and still protect the parser against all 10 payloads. These are the "several other combinations" that are taken into account into the rule and won't flag your code if you use those instead of the ones I specifically recommend here.

Reply:
@0xDC0DE thanks! Let me know whether I captured your idea in this suggestion (resolving this thread as it is marked as Outdated due to src changes above): https://github.com/returntocorp/semgrep-docs/pull/804/files#r1063303034